File: 프로그래머스/3/12971. 스티커 모으기（2）/스티커 모으기（2）.java
class Solution {
    
    public static int maxSticker(int[] arr){
        int n = arr.length;
        if (n==1) return arr[0];
        
        int[] dp = new int[n];
        dp[0] = arr[0];
        dp[1] = Math.max(arr[0], arr[1]);
        
        for (int i=2; i<n; i++){
            dp[i] = Math.max(dp[i-1], dp[i-2] + arr[i]);
        }
        
        return dp[n-1];
    }
    
    
    public int solution(int sticker[]) {
        int n = sticker.length;
        if (n == 1) return sticker[0];        
       
        int[] case1 = new int[n-1];
        System.arraycopy(sticker, 0, case1, 0, n-1);
        int[] case2 = new int[n-1];
        System.arraycopy(sticker, 1, case2, 0, n-1);
        
        return Math.max(maxSticker(case1), maxSticker(case2));
    }
}

File: 프로그래머스/3/12971. 스티커 모으기（2）/스티커 모으기（2）.py
def solution(sticker):
    
    N = len(sticker)
    
    # 예외처리 : 1개일 때 
    if N == 1 : 
        return sticker[0]
    
    # DP함수 정의
    def max_sticker(arr):
        dp = [0] * len(arr)
        dp[0] = arr[0]
        if len(arr) > 1 :
            dp[1] = max(arr[0], arr[1])
        
        for i in range(2, len(arr)):
            dp[i] = max(dp[i-1], dp[i-2] + arr[i])
        
        return dp[-1]
    
    # case1 : 첫번째 스티커 포함
    case1 = max_sticker(sticker[:-1])
    # case2 : 마지막 스티커 포함
    case2 = max_sticker(sticker[1:])
    
    return max(case1, case2)